File: coding-gauntlet/gate4-pocket-agora/Sources/PocketAgora/Protocol.swift
// Protocol.swift — FROZEN. The Swift mirror of Gate 2's protocol.ts (g2.1) + Gate 3's plugin
// surface. This is the contract the iOS app INHERITS: it decodes exactly what the Gate 2/3
// server emits and encodes exactly what the server accepts. Keep in lockstep with protocol.ts.
import Foundation

public let PROTOCOL_VERSION = "g2.1"

// ── Canonical message schema (mirrors schema.ts) ──────────────────────────────
public enum AuthorKind: String, Codable { case human, model }
public enum Role: String, Codable { case user, assistant, system }

public struct Author: Codable, Hashable {
    public var id: String
    public var kind: AuthorKind
    public var name: String?
    public var handle: String?
    public init(id: String, kind: AuthorKind, name: String? = nil, handle: String? = nil) {
        self.id = id; self.kind = kind; self.name = name; self.handle = handle
    }
}

public struct Message: Codable, Identifiable, Hashable {
    public var id: String
    public var room: String
    public var author: Author
    public var role: Role
    public var content: String
    public var ts: Double   // epoch ms (JSON number)
    public init(id: String, room: String, author: Author, role: Role, content: String, ts: Double) {
        self.id = id; self.room = room; self.author = author; self.role = role; self.content = content; self.ts = ts
    }
    public var date: Date { Date(timeIntervalSince1970: ts / 1000.0) }
}

// ── Entities ──────────────────────────────────────────────────────────────────
public struct Account: Codable, Identifiable, Hashable {
    public var id: String; public var handle: String; public var kind: AuthorKind; public var createdAt: Double?
}
public struct Room: Codable, Identifiable, Hashable {
    public var id: String; public var name: String; public var createdAt: Double?
}
public struct Membership: Codable, Hashable { public var roomId: String; public var accountId: String; public var joinedAt: Double? }

// ── REST DTOs ─────────────────────────────────────────────────────────────────
public struct AuthRequest: Codable { public var handle: String; public var password: String
    public init(handle: String, password: String){ self.handle = handle; self.password = password } }
public struct AuthResult: Codable { public var accountId: String; public var handle: String; public var token: String }
public struct CreateRoomReq: Codable { public var name: String; public init(name: String){ self.name = name } }
public struct AddAgentReq: Codable {
    public var handle: String; public var endpoint: String; public var model: String?; public var persona: String?
    public init(handle: String, endpoint: String, model: String? = nil, persona: String? = nil){
        self.handle = handle; self.endpoint = endpoint; self.model = model; self.persona = persona }
}
public struct SendReq: Codable { public var content: String; public init(content: String){ self.content = content } }
public struct APIError: Codable, Error { public var error: String; public var code: Int }

// ── WebSocket frames (ws://host/ws?token=…) ───────────────────────────────────
// Client -> server
public enum ClientFrame: Encodable {
    case subscribe(roomId: String)
    case send(roomId: String, content: String)
    case ping
    public func encode(to enc: Encoder) throws {
        var c = enc.container(keyedBy: K.self)
        switch self {
        case .subscribe(let r): try c.encode("subscribe", forKey: .type); try c.encode(r, forKey: .roomId)
        case .send(let r, let content): try c.encode("send", forKey: .type); try c.encode(r, forKey: .roomId); try c.encode(content, forKey: .content)
        case .ping: try c.encode("ping", forKey: .type)
        }
    }
    enum K: String, CodingKey { case type, roomId, content }
}
// Server -> client
public enum ServerFrame: Decodable {
    case message(Message)
    case ack(ref: String)
    case presence(roomId: String, accountId: String, online: Bool)
    case error(error: String, code: Int)
    case pong
    case unknown(String)
    enum K: String, CodingKey { case type, message, ref, roomId, accountId, online, error, code }
    public init(from dec: Decoder) throws {
        let c = try dec.container(keyedBy: K.self)
        switch try c.decode(String.self, forKey: .type) {
        case "message": self = .message(try c.decode(Message.self, forKey: .message))
        case "ack": self = .ack(ref: (try? c.decode(String.self, forKey: .ref)) ?? "")
        case "presence": self = .presence(roomId: try c.decode(String.self, forKey: .roomId),
                                          accountId: try c.decode(String.self, forKey: .accountId),
                                          online: (try? c.decode(Bool.self, forKey: .online)) ?? false)
        case "error": self = .error(error: try c.decode(String.self, forKey: .error), code: (try? c.decode(Int.self, forKey: .code)) ?? 0)
        case "pong": self = .pong
        case let t: self = .unknown(t)
        }
    }
}

// ── Endpoint helpers (the inherited REST/WS surface) ──────────────────────────
public enum API {
    public static func rest(_ base: String, _ path: String) -> URL { URL(string: base.hasSuffix("/") ? base + String(path.dropFirst()) : base + path)! }
    public static func ws(_ base: String, token: String) -> URL {
        let wsBase = base.replacingOccurrences(of: "http", with: "ws")
        return URL(string: wsBase + "/ws?token=" + (token.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? token))!
    }
    public static let register = "/auth/register", login = "/auth/login", rooms = "/rooms"
    public static func roomMessages(_ id: String) -> String { "/rooms/\(id)/messages" }
    public static func roomJoin(_ id: String) -> String { "/rooms/\(id)/join" }
    public static func roomAgents(_ id: String) -> String { "/rooms/\(id)/agents" }
}

File: Faust/gate2-group-server/src/protocol.ts
// src/protocol.ts — FROZEN Gate 2 wire contract.
// This is THE API the Gate 4 iOS app (and the Gate 2/3 web client) consume verbatim:
// entities, REST DTOs, the WebSocket envelope unions, and the orchestration interfaces.
// Reuses the canonical Message/Author from schema.ts. Do not change shapes without a
// version bump — Gates 3 and 4 are bound to it.
import type { Message, Author, AuthorKind } from "./schema.js";
export type { Message, Author, AuthorKind };

export const PROTOCOL_VERSION = "g2.1" as const;

// ── Entities ──────────────────────────────────────────────────────────────────

export interface Account {
  id: string; // "u:joe" (human) or "m:critic-glm" (model) — matches Author.id
  handle: string; // unique login/display handle, e.g. "joe"
  kind: AuthorKind; // human | model
  isAdmin?: boolean; // true if this account has admin privileges
  createdAt: number; // epoch ms
}

/** A model-bound account: an LLM endpoint dressed as a first-class room participant. */
export interface AgentAccount extends Account {
  kind: "model";
  endpoint: string; // llama-server base, e.g. http://node3090.home.arpa:8642
  model?: string; // optional model id passed through
  persona?: string; // optional system prompt / role (e.g. "cross-family critic")
  maxTokens?: number; // max tokens for model responses (default 2048)
}

export interface Room {
  id: string;
  name: string;
  createdAt: number;
}

export interface Membership {
  roomId: string;
  accountId: string;
  joinedAt: number;
}

// ── REST (auth = Bearer token in Authorization header unless noted) ────────────
// POST /auth/register   {handle, password}            -> AuthResult            (no auth)
// POST /auth/login      {handle, password}            -> AuthResult            (no auth)
// GET  /rooms                                          -> Room[]
// POST /rooms           {name}                         -> Room
// POST /rooms/:id/join                                 -> Membership
// POST /rooms/:id/agents {handle, endpoint, model?, persona?, maxTokens?} -> AgentAccount  (adds a model account to the room)
// GET  /rooms/:id/messages?before?&limit?              -> Message[]            (must be a member -> else 403)
// POST /rooms/:id/messages {content}                   -> Message             (optional REST send; primary path is WS)
// Errors: 401 (no/invalid token), 403 (not a member), 404, 400.

export interface AuthResult {
  accountId: string;
  handle: string;
  token: string;
  isAdmin: boolean;
}
export interface CreateRoomReq { name: string; }
export interface AddAgentReq {
  handle: string;
  endpoint: string;
  model?: string;
  persona?: string;
  maxTokens?: number;
}
export interface SendReq { content: string; }
export interface ApiError { error: string; code: number; }

// ── WebSocket envelopes (connect: ws://host/ws?token=…) ───────────────────────

export type ClientFrame =
  | { type: "subscribe"; roomId: string } // start receiving room events
  | { type: "send"; roomId: string; content: string } // post a message
  | { type: "ping" };

export type ServerFrame =
  | { type: "message"; message: Message } // a new message in a subscribed room
  | { type: "ack"; ref: string } // server accepted a client send (ref = client-supplied or message id)
  | { type: "presence"; roomId: string; accountId: string; online: boolean }
  | { type: "error"; error: string; code: number }
  | { type: "pong" };

// ── Orchestration (locked: mention-+-reply behind a swappable SpeakerPolicy) ──

export interface SpeakerContext {
  room: Room;
  message: Message; // the message that just landed
  agents: AgentAccount[]; // model accounts present in the room
  turnBudgetRemaining: number; // fan-out budget left for THIS origin turn (0 => must stop)
}

/**
 * Decides which model agents (if any) should respond to `message`.
 * The DEFAULT and only Gate-2 policy is mention-+-reply (see MentionReplyPolicy):
 * an agent is selected only if it is @mentioned or directly replied-to, it never
 * selects the message's own author, and selection is empty once the budget is 0.
 * Autonomous-cadence / director policies are later implementations of this SAME
 * interface — not rewrites.
 */
export interface SpeakerPolicy {
  readonly name: string;
  selectResponders(ctx: SpeakerContext): AgentAccount[];
}

export interface ModelClient {
  complete(req: {
    endpoint: string;
    model?: string;
    messages: Message[];
    signal?: AbortSignal;
    maxTokens?: number;
  }): Promise<string>;
}

// Default per-human-message fan-out budget (termination guarantee). A human message
// opens a turn with this many agent responses; model messages never open a new turn,
// so any model<->model chain draws down the SAME budget and is guaranteed to halt.
export const DEFAULT_TURN_BUDGET = 6;
